### ISSUE TITLE
Preliminary Review

### ISSUE THREAD

**Issue by tmyers273** [tmyers273/laravel-stripe-billing#5]
Hi Denis!

First thoughts - this looks great. I'm not seeing any holes anywhere and it looks like it will cover all the immediate use cases I can think of.

A few questions for you to help me better understand typical use cases.

**1. Creating Plans**

```php
// Create the plans
$bronzePlan= Plan::create([
    'description' => 'Bronze Plan',
    'name' => 'bronze',
    'is_free' => false,
]);

$silverPlan= Plan::create([
    'description' => 'SilverPlan',
    'name' => 'silver',
    'is_free' => false,
]);

$goldPlan= Plan::create([
    'description' => 'Gold Plan',
    'name' => 'gold',
    'is_free' => false,
]);

// Create the PricingPlan
$bronzeMonthly = PricingPlan::create([
    'plan_id' => $bronzePlan->id, // is this correct?
    'description' => 'Monthly Bronze Plan',
    'name' => 'bronze_monthly_50.00',
    'interval' => 'month',
    'stripe_plan_id' => 'bronze_monthly', // this needs to be created in Stripe first, right?
    'price' => 5000,
    'active' => true,
]);

$bronzeYearly = PricingPlan::create([
    'plan_id' => $bronzePlan->id, // is this correct?
    'description' => 'Yearly Bronze Plan',
    'name' => 'bronze_yearly_400.00',
    'interval' => 'year',
    'stripe_plan_id' => 'bronze_yearly', // this needs to be created in Stripe first, right?
    'price' => 40000,
    'active' => true,
]);

$silverMonthly = PricingPlan::create([
    'plan_id' => $silverPlan->id, // is this correct?
    'description' => 'Monthly Silver Plan',
    'name' => 'silver_monthly_60.00',
    'interval' => 'month',
    'stripe_plan_id' => 'silver_monthly', // this needs to be created in Stripe first, right?
    'price' => 6000,
    'active' => true,
]);

// etc
```

Does that look right?

Do the Stripe plan ids need to be manually created?

**2. Assigning users to Plans**
```php
// Basic plan subscription
$user = App\User::find(1);
$pricingPlan = TMyers\StripeBilling\Models\Plan::findByName('bronze_monthly_50.00')
$user->subscribeTo($pricingPlan );

// If called again, will throw an AlreadySubscribed exception
$user->subscribeTo($pricingPlan );
```

Look right so far?

**3. Upgrading Plans**
```php
// Basic plan subscription
$user = App\User::create([ // user creation stuff ]);
$pricingPlan = TMyers\StripeBilling\Models\Plan::findByName('bronze_monthly_50.00')
$user->subscribeTo($pricingPlan );

// To upgrade
$user->subscribeTo($pricingPlan );
$subscription = $user->activeSubscriptions()[0]; // get the first active subscription
// would be nice to have a helper method to get the first
$newPlan = TMyers\StripeBilling\Models\Plan::findByName('silver_monthly_60.00')
$subscription->changeTo($newPlan);
```

Does that look right?

**4. Downgrading Plans**

Same as upgrading, right?

**5. Canceling Plans**

```php
// Basic plan subscription
$user = App\User::create([ // user creation stuff ]);
$pricingPlan = TMyers\StripeBilling\Models\Plan::findByName('bronze_monthly_50.00')
$user->subscribeTo($pricingPlan );

// To cancel
$user->subscribeTo($pricingPlan );
$subscription = $user->activeSubscriptions()[0]; // get the first active subscription
$subscription->cancelAtPeriodEnd();
```

**6. Resuming Subscriptions**

Am I correct in understanding that, due to Stripe limitations, subscriptions can only be resumed if the `cancelAtPeriodEnd()` method is used and it is before the period end date?

If a user wants to resume either after the period end date OR if their subscription was cancelled with `cancelNow()` then a new `Subscription` needs to be created, right?

**Comment by denismitr** [comment]
Hi, Tom!

> 1. Creating Plans

Correct. `plan_id` on `pricing_plans` is a nullable field which can be omitted if only plans like `monthly` `yearly` are required with no access priviledge differences. `stripe_plan_id` should come from Stripe. I was thinking about adding something like a `persist()` method to the `PricingPlan` so it would create a Stripe plan automatically. But even in the test mode this cannot be run twice, so creating Stripe plans should probably be done manually. 

> 2. Assigning users to Plans

Almost correct: `TMyers\StripeBilling\Models\Plan::findByName('bronze_monthly_50.00')` =>  ` TMyers\StripeBilling\Models\PricingPlan::findByName('bronze_monthly_50.00')`

User can only subscribe to the `PricingPlan`. It basically replicates the `Stripe` schema only with slightly different names. Stripe forces every `Plan` (`PricingPlan` in our impl) to belong to a `Product` (parent `Plan` in out implementation). However when we check user subscription - the `isSubscribedTo` method accepts a parent `Plan` or a `PricingPlan` or just a string with a name of either one. This is what happens in the `SubscriptionMiddleware` which accepts optional string parameters like `subscription:basic,pro`. Without parameters it only checks for an active subscription.

> 3. Upgrading Plans

Totally agree - helper method to retrieve first active subscription is necessary.

> 4. Downgrading 

I have some concerns here, because User may have some DB relationships that are allowed only for say `Gold` users. Like more related files or suppliers or team members, etc.. In case of downgrade this can create problems. Of course, it is not the package responsibility to check if user is allowed to switch to this ot that plan but I'm thinking whether the package could help in some way to verify the ability to downgrade.

> 5 - 6 Cancelling and Resuming

Yes, subscription can be resumed only after `cancelAtPeriodEnd()` was applied. After `cancelNow` a new subscription must be created.
